[SIL] NFC: Simplify SILVTable and save 8 bytes per SILVTable

davezarzycki opened the pull request:
**EDITED**

We are not using the primary benefits of an intrusive list, namely the ability to insert or remove from the middle of the list, so let's switch to a plain vector. This also avoids linked-list pointer chasing.

**OLD TEXT**

I noticed that vtables were being tracked redundantly by SILModule and nothing seems to be using the intrusive list aspect of SILVTable, so I tried just converging everything onto the DenseMap. This almost works, but then the destructor for SILModule blows up when it finds a function with a non-zero reference count. This surprised me.

Am I missing something obvious?

eeckstein reviewed (dismissed):
This makes the compiler non-deterministic.
For example, when iterating over the vtables for serialization.

The general rule is: never iterate over a map/set of pointers, except it can be 100% proven that the iteration order does not effect any compilation result.

davezarzycki commented:
Ya, I just figured out why the ref count was non-zero and now the test suite is blowing up and now the test suite is blowing up as you predicted. Nevertheless, sort() can solve this, no? Or is important that vtables be emitted in source order?

eeckstein commented:
It doesn't necessarily be in source order. But it needs to be deterministic.
'sort' cannot solve this.

davezarzycki commented:
So why do we have `-emit-sorted-sil` then?

eeckstein commented:
-emit-sorted-sil is to get some more consistent output in textual SIL, but it's not to fix non-determinism in the compiler.

What would be the sort predicate for vtables?

Beside that, providing an iteration API for something which is a non-deterministic map under the hood, is extremely dangerous. It's very easy to misuse this API and introduce non-determinism.

davezarzycki commented:
Fair enough. I'm observing your point now in the test suite. In any case, we can still save memory and avoid pointer chasing if we don't use an intrusive list, so I'll simplify this patch.

davezarzycki committed:
[SIL] NFC: Simplify SILVTable and save 8 bytes per SILVTable

We were not using the primary benefits of an intrusive list, namely the
ability to insert or remove from the middle of the list, so let's switch
to a plain vector. This also avoids linked-list pointer chasing.

davezarzycki commented:
@swift-ci please test

eeckstein commented on include/swift/SIL/SILModule.h:
```diff
@@ -406,17 +406,17 @@ class SILModule {
+  vtable_iterator vtable_end() { return getVTableList().end(); }
+  vtable_const_iterator vtable_begin() const { return getVTableList().begin(); }
+  vtable_const_iterator vtable_end() const { return getVTableList().end(); }
   iterator_range<vtable_iterator> getVTables() {
```
I don't think we still need getVTableList() and getVTables().
I would keep
```
  VTableListType getVTables() { return llvm::ArrayRef<SILVTable*>(vtables); }
```

jckarter commented on include/swift/SIL/SILModule.h:
```diff
@@ -178,7 +178,7 @@ class SILModule {
 
   /// The list of SILVTables in the module.
-  VTableListType vtables;
+  std::vector<SILVTable*> vtables;
```
Maybe we could use an `llvm::MapVector` here, so that we get both stable order and efficient lookup?

davezarzycki replied:
I just looked at the `llvm::MapVector` implementation. It's less time and space efficient than what this pull request is proposing (in short: "map of keys to indices into a vector of key/value pairs"). It's also less convenient because the `std::vector` is key/value pairs when all we need/want is an array of values.

jckarter replied:
Alright, sounds good.

davezarzycki commented:
@eeckstein – I've made the getVTableList() -> getVTables() change you requested.

@swift-ci please test

eeckstein reviewed (approved):
lgtm!

davezarzycki commented:
Thanks. I'm going to hold out on merging this while I make a last ditch effort to get the build czars to fix last nights erroneous merge "correctly".

davezarzycki commented:
@swift-ci please test

davezarzycki commented:
@swift-ci please test
